src/types.rs

```diff
@@ -90,12 +102,102 @@ impl Mul<usize> for BlockNum {
 pub struct TokenAmount(pub u64);
 
+/// Token - The identifier for a token contract
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Token(pub Address);
 
+// TODO: Rename price to bounty
+
+/// DealProposal - A proposal for a deal
+/// This is the data that is submitted to the Ethereum contract to create a deal
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DealProposal {
+    /// The address of the party to propose the deal to
+    pub executor_address: Address,
+    /// The length of the deal in blocks
+    pub deal_length_in_blocks: BlockNum,
+    /// The frequency with which to submit proofs to chain
+    pub proof_frequency_in_blocks: BlockNum,
+    /// The amount of tokens to pay to the executor
+    pub price: TokenAmount,
+    /// The amount of collateral the executor must post
+    pub collateral: TokenAmount,
+    /// The token to use for payment
+    pub erc20_token_denomination: Token,
+    /// The File size of the data to be stored
+    pub file_size: u64,
+    // TODO: Change these to the correct types
+    /// The CID of the data to be stored
+    // #[serde(serialize_with = "serialize_cid", deserialize_with = "deserialize_cid")]
+    pub ipfs_file_cid: String,
+    /// The blake3 hash of the data to be stored
+    // #[serde(serialize_with = "serialize_hash", deserialize_with = "deserialize_hash")]
+    pub blake3_checksum: String,
+}
+
+impl Tokenize for DealProposal {
+    fn into_tokens(self) -> Vec<ethers::abi::Token> {
+        vec![
+            Ad(self.executor_address),
+            Uint(U256::from(self.deal_length_in_blocks.0)),
+            Uint(U256::from(self.proof_frequency_in_blocks.0)),
+            Uint(U256::from(self.price.0)),
+            Uint(U256::from(self.collateral.0)),
+            Ad(self.erc20_token_denomination.0),
+            Uint(U256::from(self.file_size)),
+            Str(self.ipfs_file_cid),
+            Str(self.blake3_checksum),
+        ]
+    }
+}
+
+/// DealProposalBuilder - A builder for a deal proposal
+/// This struct handles Configuring and Building a DealProposal
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct DealProposalBuilder {
+    /// The address of the executor to to propose the deal to, as a string
+    pub executor_address: String,
+    /// The length of the deal in blocks, as an int
+    pub deal_length_in_blocks: u64,
+    /// The frequency of proofs to be submitted, as an int
+    pub proof_frequency_in_blocks: u64,
+    /// The amount of tokens to be paid to the executor per TiB, as a float
+    pub price_per_tib: f64,
+    /// The amount of tokens in collateral the executor must provide per TiB, as a float
+    pub collateral_per_tib: f64,
+    /// The Address of the token to use as denominator for the price and collateral, as a string
+    pub erc20_token_denomination: String,
+}
+
+// TODO (al): Don't we need to add executor / creator address to this?
```

Comment:
yes you are correct. this needs to be done in solidity as well. @jonahkaye @xBalbinus